src/VirtoCommerce.NotificationsModule.Data/Repositories/NotificationRepository.cs

```diff
@@ -32,29 +32,22 @@ public async Task<NotificationEntity[]> GetByIdsAsync(string[] ids, string respo
             {
                 result = await Notifications.Where(x => ids.Contains(x.Id)).OrderBy(x => x.Type).ToArrayAsync();
                 ids = result.Select(x => x.Id).ToArray();
-
-                var notificaionResponseGroup = EnumUtility.SafeParseFlags(responseGroup, NotificationResponseGroup.Full);
-
-                if (ids.Any())
-                {
-                    if (notificaionResponseGroup.HasFlag(NotificationResponseGroup.WithTemplates))
-                    {
-                        await DbContext.Set<NotificationTemplateEntity>().Where(t => ids.Contains(t.NotificationId)).LoadAsync();
-                    }
-                    if (notificaionResponseGroup.HasFlag(NotificationResponseGroup.WithAttachments))
-                    {
-                        await DbContext.Set<EmailAttachmentEntity>().Where(t => ids.Contains(t.NotificationId)).LoadAsync();
-                    }
-                    if (notificaionResponseGroup.HasFlag(NotificationResponseGroup.WithRecipients))
-                    {
-                        await DbContext.Set<NotificationEmailRecipientEntity>().Where(t => ids.Contains(t.NotificationId)).LoadAsync();
-                    }
-                }
+                await GetNestedEntitiesAsync(ids, responseGroup);
             }
             return result;
         }
 
+        public async Task<NotificationEntity[]> GetByNotificationsAsync(Notification[] notifications, string responseGroup)
+        {
+            //https://github.com/aspnet/EntityFrameworkCore/blob/release/2.2/src/EFCore.Specification.Tests/Query/SimpleQueryTestBase.ResultOperators.cs#L1333
+            //TODO: check on ef core 3.1
+            var notificationKeys = notifications.Select(x => new { Key1 = x.Type, Key2 = x.TenantIdentity.Id, Key3 = x.TenantIdentity.Type }).ToArray();
```

Comment:
Probably null reference exception when TenantIdentity = null. This can happen if notification is passed as a parameter to API .
For me, this code, based on closure and anonymous types, looks tricky :)

Reply:
doesn't throw, because there is no null in Notification
`public TenantIdentity TenantIdentity { get; set; } = TenantIdentity.Empty;`

Reply:
it doesn't work as composite key, because TenantId and TenantType could be null.  found case in EFCore.Specification.Tests, look at [the test](https://github.com/aspnet/EntityFrameworkCore/blob/release/2.2/src/EFCore.Specification.Tests/Query/SimpleQueryTestBase.ResultOperators.cs#L1333)